uefi/src/proto/shell_params.rs

```diff
@@ -0,0 +1,45 @@
+//! `ShellParams` protocol
+
+use crate::proto::unsafe_protocol;
+use crate::{data_types, Char16};
+use core::slice::from_raw_parts;
+use uefi_raw::protocol::shell_params::ShellParametersProtocol;
+
+#[cfg(feature = "alloc")]
+use crate::CStr16;
+#[cfg(feature = "alloc")]
+use alloc::string::String;
+#[cfg(feature = "alloc")]
+use alloc::string::ToString;
+
+/// The ShellParameters protocol.
+#[derive(Debug)]
+#[repr(transparent)]
+#[unsafe_protocol(ShellParametersProtocol::GUID)]
+pub struct ShellParameters(ShellParametersProtocol);
+
+impl ShellParameters {
+    /// Get an iterator of the shell parameter arguments
+    #[cfg(feature = "alloc")]
+    pub fn get_args(&self) -> impl Iterator<Item = String> {
+        unsafe {
+            from_raw_parts(
+                self.0.argv.cast::<*const data_types::chars::Char16>(),
```

Comment:
I don't know why but the compiler thinks `argv` is `*const *const u16` instead of `*const *const Char16`.

Reply:
The type of `Char16` is defined differently in `uefi-raw` vs `uefi`. `uefi-raw` just uses a `u16` so any value is accepted, whereas in `uefi` we use the `ucs2` crate that places some restrictions on the valid character values. (At some point I'd like to improve this situation, see https://github.com/rust-osdev/uefi-rs/pull/809#issuecomment-1551623475 for more background.)

All of which is to say, this is expected :)